File: 146.lru-cache.go
package main

import "fmt"

func main() {
	cache := Constructor(2)
}

type Node struct {
	Key        int
	Val        int
	prev, next *Node
}

type DList struct {
	head, tail *Node
}

func (dl *DList) appendLast(n *Node) {
	pr := dl.tail.prev
	pr.next = n
	n.prev = pr
	n.next = dl.tail
	dl.tail.prev = n
}

func (dl *DList) remove(n *Node) {
	n.prev.next = n.next
	n.next.prev = n.prev
}

func (dl *DList) tranverse() {
	p := dl.head.next
	for p != dl.tail {
		fmt.Print(p.Val, "->")
		p = p.next
	}
}

type LRUCache struct {
	table    map[int]*Node
	dl       *DList
	capacity int
}

func Constructor(capacity int) LRUCache {
	n1, n2 := &Node{}, &Node{}
	n1.next = n2
	n2.prev = n1
	return LRUCache{make(map[int]*Node), &DList{head: n1, tail: n2}, capacity}
}

func (this *LRUCache) Get(key int) int {
	if n, ok := this.table[key]; ok {
		this.dl.remove(n)
		this.dl.appendLast(n)
		return n.Val
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if n, ok := this.table[key]; ok {
		this.dl.remove(n)
		this.dl.appendLast(n)
		n.Val = value
	} else {
		if len(this.table) == this.capacity {
			n := this.dl.head.next
			if n == this.dl.tail {
				return
			}
			this.dl.remove(n)
			delete(this.table, n.Key)
		}

		newNode := &Node{Key: key, Val: value}
		this.dl.appendLast(newNode)
		this.table[key] = newNode
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: 146.lru-cache.java
class LRUCache {

    private Integer capacity;
    private HashMap<Integer, LinkedNode> m;
    private LinkedNode head;
    private LinkedNode end;

    static class LinkedNode {
        public Integer key;
        public Integer val;
        public LinkedNode prev;
        public LinkedNode next;

        public LinkedNode(Integer key, Integer val) {
            this.key = key;
            this.val = val;
            this.prev = null;
            this.next = null;
        }
    }

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.m = new HashMap<>();
        this.head = new LinkedNode(-1, -1);
        this.end = new LinkedNode(-1, -1);
        this.head.next = this.end;
        this.end.prev = this.head;
    }

    public int get(int key) {
        // System.out.printf("get key:%d\n", key);

        if (!this.m.containsKey(key)) {
            return -1;
        }
        this.moveFront(key);
        // this.iterate();
        return this.m.get(key).val;
    }

    public void put(int key, int value) {
        // System.out.printf("put key:%d, value:%d\n", key, value);
        if (this.m.containsKey(key)) {
            this.m.get(key).val = value;
            this.moveFront(key);
            return;
        }
        if (this.capacity.equals(this.m.size())) {
            LinkedNode endPrev = this.end.prev;
            LinkedNode end2ndPrev = endPrev.prev;
            end2ndPrev.next = this.end;
            this.end.prev = end2ndPrev;

            // System.out.printf("remove val: %d\n", endPrev.key);
            this.m.remove(endPrev.key);
        }
        LinkedNode newNode = new LinkedNode(key, value);
        LinkedNode first = this.head.next;
        newNode.prev = this.head;
        newNode.next = first;
        first.prev = newNode;
        this.head.next = newNode;
        this.m.put(key, newNode);
        // this.iterate();
    }

    private void moveFront(Integer key) {
        // System.out.printf("move %d to front\n", key);
        LinkedNode n = this.m.get(key);
        LinkedNode prev = n.prev;
        LinkedNode next = n.next;
        prev.next = next;
        next.prev = prev;
        this.head.next.prev = n;
        n.next = this.head.next;
        n.prev = this.head;
        this.head.next = n;
    }

    private void iterate() {
        LinkedNode i = this.head;
        while (i != null) {
            System.out.printf("%d ", i.key);
            i = i.next;
        }
        System.out.printf("\n");
        i = this.end;
        while (i != null) {
            System.out.printf("%d ", i.key);
            i = i.prev;
        }
        System.out.printf("\n");
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */